<a href="https://colab.research.google.com/github/HengamehSM/Kaggle/blob/main/Intro_to_Programming_Exercise_Arithmetic_and_Variables.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:

# IMPORTANT: RUN THIS CELL IN ORDER TO IMPORT YOUR KAGGLE DATA SOURCES
# TO THE CORRECT LOCATION (/kaggle/input) IN YOUR NOTEBOOK,
# THEN FEEL FREE TO DELETE THIS CELL.
# NOTE: THIS NOTEBOOK ENVIRONMENT DIFFERS FROM KAGGLE'S PYTHON
# ENVIRONMENT SO THERE MAY BE MISSING LIBRARIES USED BY YOUR
# NOTEBOOK.

import os
import sys
from tempfile import NamedTemporaryFile
from urllib.request import urlopen
from urllib.parse import unquote, urlparse
from urllib.error import HTTPError
from zipfile import ZipFile
import tarfile
import shutil

CHUNK_SIZE = 40960
DATA_SOURCE_MAPPING = 'titanic:https%3A%2F%2Fstorage.googleapis.com%2Fkaggle-competitions-data%2Fkaggle-v2%2F3136%2F26502%2Fbundle%2Farchive.zip%3FX-Goog-Algorithm%3DGOOG4-RSA-SHA256%26X-Goog-Credential%3Dgcp-kaggle-com%2540kaggle-161607.iam.gserviceaccount.com%252F20240821%252Fauto%252Fstorage%252Fgoog4_request%26X-Goog-Date%3D20240821T213103Z%26X-Goog-Expires%3D259200%26X-Goog-SignedHeaders%3Dhost%26X-Goog-Signature%3D54ba6ce1019f8c0f47b72fa73c9e13d3c132ef871b1dcf4639129b5dd5d5aeb137865dec6918e39a92503c8c32184011ed3fddebe9f697c2ff89204c6dd357fdd9b1229f255d0acae913d026da1f81a522ffbf7eb396f11a7d10addd9b81cae0cfe690f6849bcd5f7034881d0cec70990ebf14874febff8e194bcbc096a3242eced8d19097fe6c396d0657ec17c60788bd1ae9b12688995939a406355a6ab31f267c6042487ca2d055fa04d10ca04177e8241d8c3b5f21f9e4ccbcb35a23542c526f028a1ac62271f0a373f1c39a8135a0b8e0d2f8327a50869a9cfa8b33b1fce57bde4e755f14e11021b5ce85dfb4276a3b66e4d186ad04c331b4d90f9512d0'

KAGGLE_INPUT_PATH='/kaggle/input'
KAGGLE_WORKING_PATH='/kaggle/working'
KAGGLE_SYMLINK='kaggle'

!umount /kaggle/input/ 2> /dev/null
shutil.rmtree('/kaggle/input', ignore_errors=True)
os.makedirs(KAGGLE_INPUT_PATH, 0o777, exist_ok=True)
os.makedirs(KAGGLE_WORKING_PATH, 0o777, exist_ok=True)

try:
  os.symlink(KAGGLE_INPUT_PATH, os.path.join("..", 'input'), target_is_directory=True)
except FileExistsError:
  pass
try:
  os.symlink(KAGGLE_WORKING_PATH, os.path.join("..", 'working'), target_is_directory=True)
except FileExistsError:
  pass

for data_source_mapping in DATA_SOURCE_MAPPING.split(','):
    directory, download_url_encoded = data_source_mapping.split(':')
    download_url = unquote(download_url_encoded)
    filename = urlparse(download_url).path
    destination_path = os.path.join(KAGGLE_INPUT_PATH, directory)
    try:
        with urlopen(download_url) as fileres, NamedTemporaryFile() as tfile:
            total_length = fileres.headers['content-length']
            print(f'Downloading {directory}, {total_length} bytes compressed')
            dl = 0
            data = fileres.read(CHUNK_SIZE)
            while len(data) > 0:
                dl += len(data)
                tfile.write(data)
                done = int(50 * dl / int(total_length))
                sys.stdout.write(f"\r[{'=' * done}{' ' * (50-done)}] {dl} bytes downloaded")
                sys.stdout.flush()
                data = fileres.read(CHUNK_SIZE)
            if filename.endswith('.zip'):
              with ZipFile(tfile) as zfile:
                zfile.extractall(destination_path)
            else:
              with tarfile.open(tfile.name) as tarfile:
                tarfile.extractall(destination_path)
            print(f'\nDownloaded and uncompressed: {directory}')
    except HTTPError as e:
        print(f'Failed to load (likely expired) {download_url} to path {destination_path}')
        continue
    except OSError as e:
        print(f'Failed to load {download_url} to path {destination_path}')
        continue

print('Data source import complete.')


**This notebook is an exercise in the [Intro to Programming](https://www.kaggle.com/learn/intro-to-programming) course.  You can reference the tutorial at [this link](https://www.kaggle.com/alexisbcook/arithmetic-and-variables).**

---


This exercise will get you started with running your own code.  

# Set up the notebook

To begin, run the code in the next cell.
- Begin by clicking inside the code cell.  
- Click on the triangle (in the shape of a "Play button") that appears to the left of the code cell.
- If your code was run sucessfully, you will see `Setup Complete` as output below the cell.

Instead of clicking on the triangle, you can also run code by pressing Shift + Enter on your keyboard.  Try this now!  Nothing bad will happen if you run the code more than once.

In [ ]:
# Set up the exercise
from learntools.core import binder
binder.bind(globals())
from learntools.intro_to_programming.ex1 import *
print('Setup complete.')

Setup complete.


The code above sets up the notebook so that it can check your answers in this exercise.  You should never modify this code.  (Otherwise, the notebook won't be able to verify that you have successfully completed the exercise.)

After finishing all of the questions below, you'll see the exercise marked as complete on the [course page](http://www.kaggle.com/learn/intro-to-programming).  Once you complete all of the lessons, you'll get a course completion certificate!

# Question 1

Next, you will run some code from the tutorial, so you can see how it works for yourself.  Run the next code cell without changes.

In [ ]:
print("Hello, world!")

# DO NOT REMOVE: Mark this question as completed
q1.check()

Hello, world!


<IPython.core.display.Javascript object>

<span style="color:#33cc33">If you see 'Hello, world!' above, You have successfully printed a message, and you're ready to move on to the next question.</span>

You just ran code to print `Hello world!`, which you should see in the output above.

The second line of code (`q1.check()`) checks your answer.  You should never modify this checking code; if you remove it, you won't get credit for completing the problem.

# Question 2

Now, you will print another message of your choosing.  To do this, change `print("Your message here!")` to use a different message.  For instance, you might like to change it to something like:
- `print("Good morning!")`
- `print("I am learning how to code :D")`

Or, you might like to see what happens if you write something like `print("3+4")`.  Does it return 7, or does it just think of `"3+4"` as just another message?

Make sure that your message is enclosed in quotation marks (`"`), and the message itself does not use quotation marks. For instance, this will throw an error: `print("She said "great job" and gave me a high-five!")` because the message contains quotation marks.  If you decide to take the Python course after completing this course, you will learn more about how to avoid this error in [Lesson 6](https://www.kaggle.com/colinmorris/strings-and-dictionaries).

Feel free to try out multiple messages!

In [ ]:
# TODO: Change the message
print("Good morning!")

# DO NOT REMOVE: Mark this question as completed
q2.check()

Good morning!


<IPython.core.display.Javascript object>

<span style="color:#33cc33">Once you have printed your own message, you're ready to move on to the next question.</span>

# Question 3

As you learned in the tutorial, a comment in Python has a pound sign (`#`) in front of it, which tells Python to ignore the text after it.

Putting a pound sign in front of a line of code will make Python ignore that code.  For instance, this line would be ignored by Python, and nothing would appear in the output:
```python
#print(1+2)
```
Removing the pound sign will make it so that you can run the code again. When we remove the pound sign in front of a line of code, we call this **uncommenting**.

In this problem, you will uncomment two lines in the code cell below and view the output:
- Remove the `#` in front of `q3.hint()`.  To avoid errors, do NOT remove the `#` in front of `# Uncomment to view hint`.  
- Next, remove the `#` in front of `q3.solution()`.

As in the previous questions, do not change the final line of code that marks your work as completed.

In [ ]:
# Uncomment to get a hint
q3.hint()

# Uncomment to view solution
q3.solution()

# DO NOT REMOVE: Check your answer
q3.check()

<IPython.core.display.Javascript object>

<span style="color:#3366cc">Hint:</span> If you're ever stuck on a question, it's a good idea to look at the hint before viewing the solution.

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> If you're still stuck on a question after viewing the hint and re-reading the tutorial, you can view the solution.  You can also view the solution after you have successfully submitted your own answer, to check if the official solution is any different (there may be more than one right answer!).

<IPython.core.display.Javascript object>

<span style="color:#33cc33">Once you have printed the hint and the solution, you're ready to move on to the next question.</span>

In the next question, and in most of the exercises in this course, you will have the option to uncomment to view hints and solutions.  Once you feel comfortable with uncommenting, continue to the next question.


# Question 4

In the tutorial, you defined several variables to calculate the total number of seconds in a year.  Run the next code cell to do the calculation here.

In [ ]:
# Create variables
num_years = 4
days_per_year = 365
hours_per_day = 24
mins_per_hour = 60
secs_per_min = 60

# Calculate number of seconds in four years
total_secs = secs_per_min * mins_per_hour * hours_per_day * days_per_year * num_years
print(total_secs)

126144000


Use the next code cell to:
- Define a variable `births_per_min` and set it to 250.  (There are on average 250 babies born each minute.)
- Define a variable `births_per_day` that contains the average number of babies born each day.  (To set the value of this variable, you should use `births_per_min` and some of the variables from the previous code cell.)

Remember you can always get a hint if you need it!

In [ ]:
# TODO: Set the value of the births_per_min variable
births_per_min = 250

# TODO: Set the value of the births_per_day variable
births_per_day = births_per_min * hours_per_day * mins_per_hour

# DO NOT REMOVE: Check your answer
q4.check()

<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct</span>

In [ ]:
# Uncomment to get a hint
q4.hint()

# Uncomment to view solution
q4.solution()

<IPython.core.display.Javascript object>

<span style="color:#3366cc">Hint:</span> How can you use the variables to calculate the number of minutes in one day?  Once you have that, you need only multiply that number by the number of births per minute.

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> 
```python
# Set the value of the births_per_min variable
births_per_min = 250

# Set the value of the births_per_day variable
births_per_day = births_per_min * mins_per_hour * hours_per_day

```

# 🌶️ Question 5

(Questions marked with a 🌶️ will be a little bit more challenging than the others!  Remember you can always get a hint or view the solution.)

The [Titanic competition](https://www.kaggle.com/c/titanic) is Kaggle's most famous data science competition. In this competition, participants are challenged to build a machine learning model that can predict whether or not passengers survived the Titanic shipwreck, based on information like age, sex, family size, and ticket number.

Run the next code cell without changes to load and preview the titanic data.

Don't worry about the details of the code for now - the end result is just that the all of the titanic data has been loaded in a variable named `titanic_data`.  (In order to learn how to write this code yourself, you can take the [Python course](https://www.kaggle.com/learn/python) and then the [Pandas course](https://www.kaggle.com/learn/pandas).)

In [ ]:
# Load the data from the titanic competition
import pandas as pd
titanic_data = pd.read_csv("../input/titanic/train.csv")

# Show the first five rows of the data
titanic_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

The data has a different row for each passenger.  

The next code cell defines and prints the values of three variables:  
- `total` = total number of passengers who boarded the ship
- `survived` = number of passengers who survived the shipwreck
- `minors` = number of passengers under 18 years of age

Run the code cell without changes.  (Don't worry about the details of how these variables are calculated for now.  You can learn more about how to calculate these values in the [Pandas course](https://www.kaggle.com/learn/pandas).)

In [ ]:
# Number of total passengers
total = len(titanic_data)
print(total)

# Number of passengers who survived
survived = (titanic_data.Survived == 1).sum()
print(survived)

# Number of passengers under 18
minors = (titanic_data.Age < 18).sum()
print(minors)

891
342
113


So,
- `total = 891` (there were 891 passengers on board the Titanic),
- `survived = 342` (342 passengers survived), and
- `minors = 113` (113 passengers were under the age of 18).

In the code cell below, replace the underlines (`____`) with code to calculate the values for two more variables:
- `survived_fraction` should be set to the fraction of passengers who survived the Titanic disaster.
- `minors_fraction` should be the fraction of passengers who were minors (under the age of 18).

For each variable, your answer should be a number between 0 and 1.  

If you need a hint or want to view the solution, you can skip to the next code cell and uncomment the appropriate lines of code (`q5.hint()` and `q5.solution()`).

In [ ]:
# TODO: Fill in the value of the survived_fraction variable
survived_fraction = survived / total

# Print the value of the variable
print(survived_fraction)

# TODO: Fill in the value of the minors_fraction variable
minors_fraction = minors / total

# Print the value of the variable
print(minors_fraction)

# DO NOT REMOVE: Check your answer
q5.check()

0.3838383838383838
0.12682379349046016


<IPython.core.display.Javascript object>

<span style="color:#33cc33">Correct</span>

In [ ]:
# Uncomment to receive a hint
q5.hint()

# Uncomment to view the solution
q5.solution()

<IPython.core.display.Javascript object>

<span style="color:#3366cc">Hint:</span> To get the fraction of people who survived the titanic, you need to divide the number of survivors by the total number of people.  Remember the variables that you can use to answer this question: `survived`, `total`, and `minors`.

<IPython.core.display.Javascript object>

<span style="color:#33cc99">Solution:</span> 
```python
# Fill in the value of the survived_fraction variable 
survived_fraction = survived/total

# Print the value of the survived_fraction variable
print(survived_fraction)

# Fill in the value of the minors_fraction variable
minors_fraction = minors/total 

# Print the value of the minors_fraction variable
print(minors_fraction)

```

# Bonus Exercise

As an **optional** next step, you're encouraged to make a submission to the Titanic competition.  
> To do this, follow the instructions in the **[notebook here](https://www.kaggle.com/alexisbcook/titanic-tutorial)**.  

The notebook is beginner-friendly and does not assume you have any experience with coding.  At the same time, you'll learn how to author your own notebooks on Kaggle, and it gives you a great idea of what you'll be able to do if you continue to learn data science!

# Keep going

Congratulations!  You have just finished the first exercise in the Intro to Programming course!  

You are now ready to learn how to **[organize your code with functions](https://www.kaggle.com/alexisbcook/functions)**.

---




*Have questions or comments? Visit the [course discussion forum](https://www.kaggle.com/learn/intro-to-programming/discussion) to chat with other learners.*